### PR TITLE
Experimental tracing module (3/3): define and expose a `instrumentHTTP` function

### PR DESCRIPTION
**Pull request by oleiade**
## About

To keep Pull Requests as small as possible and reduce the review burden, this PR is the third one in a chain of three implementing the decided API design for the experimental tracing module.

warning ✋🏻 This PR is part of a PR chain, and is based on https://github.com/grafana/k6/pull/2854

This second PR builds upon #2853 and #2854  and adds an `instrumentHTTP` function to k6/experimental/tracing.

Using this function in their k6 scripts, once, users can automatically have all their calls to the `http` module's methods (`http.{del,get,head,options,patch,post,put,request}`) automatically instrumented with tracing information, in the same fashion as described in #2854. Indeed, this function under the hood remaps the `http` module's method to use the `Client`'s methods instead.

The `instrumentHTTP` takes the same option set as the `Client` as input but does not allow for reconfiguration, nor does it allow to perform non instrumented calls. As it is the intended behavior, once called, the `http` module's methods are replaced with the instrumented ones. 

## In action

```javascript
import http from "k6/http";
import { check } from "k6";
import tracing from "k6/experimental/tracing";

// instrumentHTTP will ensure that all requests made by the http module
// will be traced. The first argument is a configuration object that
// can be used to configure the tracer.
//
// Currently supported HTTP methods are: get, post, put, patch, head,
// del, options, and request.
tracing.instrumentHTTP({
	propagator: "w3c",
});

export default () => {
	let res = http.get("http://httpbin.org/get", {
		headers: {
			"X-Example-Header": "instrumented/get",
		},
	});
	check(res, {
		"status is 200": (r) => r.status === 200,
	});
};
```

To observe this example in action, I recommend running it with the --http-debug too. That way, you'll be able to observe the Traceparent header being added (note that when using the b3 or jaeger propagators, the header is named differently.

```
go run go.k6.io/k6 run --http-debug samples/experimental/tracing.js
```

## Review

⚠️ Although this PR has >1000 lines addition, a considerable chunk of it is testing. I believe the amount of actual logic is somewhat limited ⚠️ 

The commits have been organized to be progressive and make sense as a sequence; thus, you should review them in order.

## References

Previous: #2853 #2854

### CODE REVIEW

File: js/modules/k6/experimental/tracing/tracing.go
Comment (mstoykov) [inline js/modules/k6/experimental/tracing/tracing.go:48]: Does this need to be a field ?

File: js/modules/k6/experimental/tracing/tracing.go
Comment (mstoykov) [inline js/modules/k6/experimental/tracing/tracing.go:94]: This shouldn't be needed as you are directly changign the export. And it won't work if it is imported as anythign else.
```suggestion
```

File: cmd/integration_test.go
Comment (mstoykov) [inline cmd/integration_test.go:1914]: Same comment as in the previous PR, I guess you can even make this check into a function that returns a true if it ever got a single metric that matches all the conditions.

File: js/modules/k6/experimental/tracing/tracing.go
Comment (oleiade) [reply js/modules/k6/experimental/tracing/tracing.go:48]: Probably not indeed 👍🏻

File: js/modules/k6/experimental/tracing/tracing.go
Comment (oleiade) [reply js/modules/k6/experimental/tracing/tracing.go:94]: Good catch indeed! 🙇🏻

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:214]: I would argue those are likely better as just inline constants instead of importing the whole package(not that you don't indirectly). This also means that if you somehow change it in the tracing API it won't suddenly break while the test is still passing ;)

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:274]: Can we have this as a helper function as it seems to be done multiple times.

File: cmd/tests/tracing_module_test.go
Comment (codebien) [inline cmd/tests/tracing_module_test.go:138]: an atomic should be better for avoiding any race

File: js/modules/k6/experimental/tracing/module.go
Comment (mstoykov) [inline js/modules/k6/experimental/tracing/module.go:23]: Is there a need for this field? It seems to me it is only used in one place so :shrug:

File: js/modules/k6/experimental/tracing/module.go
Comment (mstoykov) [inline js/modules/k6/experimental/tracing/module.go:92]: nitpick: this doesn't prevent someone from calling `instrumentHTTP` multiple times with different settings.

I am not certain what the solution should be here. I would argue changing the client somehow instead of stacking them. But it is probably okay if we just return error for now :shrug:

**Review by codebien — CHANGES_REQUESTED** [review]

File: js/modules/k6/experimental/tracing/module.go
Comment (oleiade) [reply js/modules/k6/experimental/tracing/module.go:92]: I agree with you, I would find it more convenient and explicit if we threw an error if calling `instrumentHTTP` a second time. We could even make the error explicit along the lines of "it looks like you're attempting to reconfigure the HTTP module's instrumentation on the fly. This is not supported. Consider using the tracing.Client instead".

File: js/modules/k6/experimental/tracing/module.go
Comment (oleiade) [reply js/modules/k6/experimental/tracing/module.go:23]: Following your comment on throwing errors when `instrumentHTTP` is called twice, I now rely on this field to check if we have already initialized a module-wide client or not 👍🏻

**Review by codebien — CHANGES_REQUESTED** [review]
The CI is failing

**Comment by oleiade** [comment]
@codebien indeed, sorry for the noise... 🙇🏻  It should be fixed now 😸 (once CI completes)

File: js/modules/k6/experimental/tracing/module.go
Comment (codebien) [inline js/modules/k6/experimental/tracing/module.go:84]: Should it be protected from data race? Could it be called concurrently? Apparently, nothing blocks the user from using it in the iteration function.

Comment (oleiade) [reply js/modules/k6/experimental/tracing/module.go:84]: I'm not sure about the potential data race; probably @mstoykov might be more able to give a certain answer on this? My current reasoning is the module is instantiated once for each VU; thus this check will access the `Client` instance only once per-vu at a time, hence, no concurrent access? I might be completely off on this though, as I don't have the VU execution model freshly loaded in my brain 🤔 

Your second point actually brings me to think that considering `instrumentHTTP` is a one-way street at the moment, we probably want to limit it to the init context indeed 🤔

Comment (oleiade) [reply js/modules/k6/experimental/tracing/module.go:88]: Indeed, thanks for the pointer 🙏🏻 I had already done it in xk6-redis too, though 😸 I've added a check, and a couple tests to cover that 🤝

Comment (mstoykov) [reply js/modules/k6/experimental/tracing/module.go:88]: > the module is instantiated once for each VU

Yes this is true . 

There is no possible race condition as it will require that this is called twice *from* js at the same time.

But by specification js is single threaded so there is no way for that to happen or it will be a bug in something else - like using a js runtime concurrently.

**Review by codebien — APPROVED** [review]
LGTM

File: cmd/tests/tracing_module_test.go
Comment (codebien) [inline cmd/tests/tracing_module_test.go:115]:
```suggestion

```
It does seem to be required

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:285]: I kind of prefer if those are not integration tests as they are more expensive and in this case there is no need for them IMO.

The functionality tested is pretty common :
1. exception in an iteration does not abort the whole test
2. an exception in the init context - does.

**Review by mstoykov — APPROVED** [review]
LGTM, but if possible I would like to move some of the integration tests to not integration ones :bow:

File: cmd/tests/tracing_module_test.go
Comment (codebien) [inline cmd/tests/tracing_module_test.go:206]:
```suggestion
```
@oleiade there are multiple lines of it

File: cmd/tests/tracing_module_test.go
Comment (oleiade) [reply cmd/tests/tracing_module_test.go:285]: Let me see if I can easily turn them into standard unit tests indeed. I've struggled with `instrumentHTTP` depending on require in the past, but with the changes with made, it might not be such an issue anymore 👍🏻

File: cmd/tests/tracing_module_test.go
Comment (oleiade) [reply cmd/tests/tracing_module_test.go:206]: Good catch! :eyes: I guess this is the downside of writing JS in a Go string, no eslint, no prettier, no checks whatsoever.

**Review by mstoykov — COMMENTED** [review]
I guess we are do not have an integration test that instrumenting the API in one file will instrtument it across teh whole k6.

This likely is the most interesting integration test, so I think it is worth doing before merging

**Comment by oleiade** [comment]
@mstoykov thanks a lot for your feedback 🙇🏻 I've replaced the relevant integration tests with unit tests. I've also added an integration test demonstrating `instrumentHTTP` instruments across the whole of k6.

File: js/modules/k6/experimental/tracing/module_test.go
Comment (mstoykov) [inline js/modules/k6/experimental/tracing/module_test.go:79]: nit: I feel like we can pull this out in a helper function like in other modules. One that returns a struct with fields of the necessary thing for the test like the modulestest.Runtime, the moduleinstance if necessary and so on.

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:250]: nit: Having asserts like this is in general an anti-pattern and the testing package will give you warnings if you ever actually fail here. 

Although maybe with assert instead of require it is fine as it won't call [FailNow](https://pkg.go.dev/testing#T.FailNow) which is the usual problem :thinking:

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:290]: None of those check that *both* requests have the traceid. Given that we already check for the one in other tests I will recommend just having 1 requests that is made *not* from the module calling `instrumentHTTP`

File: js/modules/k6/experimental/tracing/module_test.go
Comment (oleiade) [reply js/modules/k6/experimental/tracing/module_test.go:79]: I see what you mean, and you're right I'll move it in a dedicated `testSetup` struct like in other places 🤝

**Review by mstoykov — CHANGES_REQUESTED** [review]

File: cmd/tests/tracing_module_test.go
Comment (oleiade) [reply cmd/tests/tracing_module_test.go:250]: It works as I expected, though. I've just forced the test to fail at this statement, and I see no such warnings 🤔 

```
/home/oleiade/Dev/Grafana/k6/cmd/tests/tracing_module_test.go:250:
        	Error Trace:	/home/oleiade/Dev/Grafana/k6/cmd/tests/tracing_module_test.go:250
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2109
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2487
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2947
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:1991
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/asm_amd64.s:1594
        	Error:      	"00-1234567890-bf358b23435cbdd7-01" should have 55 item(s), but has 33
        	Test:       	TestTracingInstrummentHTTP_SupportsMultipleTestScripts
```

File: cmd/tests/tracing_module_test.go
Comment (oleiade) [reply cmd/tests/tracing_module_test.go:290]: You're absolutely right, that's a really good point 👍🏻

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [reply cmd/tests/tracing_module_test.go:250]: Do you get a warning if you use `require` :thinking:

Comment (oleiade) [reply cmd/tests/tracing_module_test.go:250]: Nope, same output

```
/home/oleiade/Dev/Grafana/k6/cmd/tests/tracing_module_test.go:250:
        	Error Trace:	/home/oleiade/Dev/Grafana/k6/cmd/tests/tracing_module_test.go:250
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2109
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2487
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:2947
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/server.go:1991
        	            				/home/oleiade/Dev/Grafana/k6/cmd/tests/asm_amd64.s:1594
        	Error:      	"00-1234567890-3d5a73e6627d9b8e-01" should have 55 item(s), but has 33
        	Test:       	TestTracingInstrummentHTTP_SupportsMultipleTestScripts
```

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [inline cmd/tests/tracing_module_test.go:237]: Kind of the same comment about this not checking that all of those have tracing ids.

Maybe `assertHasTraceIDMetadata` can have a number argument that is how many it expects :shrug:. 

Or to take the expected url(s) (as vararg not required argument)  and check that all of those have trace id metadata sample.

File: cmd/tests/tracing_module_test.go
Comment (mstoykov) [reply cmd/tests/tracing_module_test.go:250]: Maybe I am mistaken and it only matters if it happens after the [test functions has returned](https://go.dev/play/p/cAtYXKMIOwBc) :thinking:  or if you have sub test and use the parent one :shrug: 

Whichever - it doesn't really matter. Sorry for the noise :bow:

**Review by mstoykov — APPROVED** [review]

**Comment by mstoykov** [comment]
I have left one [comment](https://github.com/grafana/k6/pull/2855#discussion_r1087664171) that is not blocking IMO given that it is very unlikely breakage in that case (IMO)

**Comment by oleiade** [comment]
Thanks a lot for the reviews folks 🙇🏻 👏🏻 much appreciated!

For the sake of getting a sense of achievement, I'll go ahead and merge this. I think your point on the `assertHasTraceIDMetadata` is valid @mstoykov and I will try to address it in a further PR 🦕 

🚀 🚀 🚀

**Pull request merged**
